### DISCUSSION TITLE
Start button and first lap

### DISCUSSION THREAD

**Discussion by ftsaci** [micbay/RaceTimerAndController#22] [General]
Thank you for this nice timer. Please excuse me for what I am going to ask. Unfortunately, I do not have any experience with coding. Can you help me with a few small changes? Is it possible to start a quick race with the current settings when the start button is pressed on the main screen? And secondly, can we record the first trigger after the start in the standard race as a lap time?

**Comment by micbay** [comment]
I'm not 100% sure I fully understand your request, but this is my first stab at it. I do not have anything with me to test so this is just my best guess as to what you might be able to do, relatively quickly to try and get what I think you are asking.

Immediate Start - It seems you are asking to skip the staging screen and want to go straight into the pre-start countdown and race start when A, B, or C(drag) race is selected from the race type select menu. In the main code (file [RaceTimerAndController.ino](https://github.com/micbay/RaceTimerAndController/blob/master/RaceTimerAndController.ino#L2305)), the `StartRaceMenu` case (line 2139) defines what happens from this menu page. If you want to skip the `Staging` state, and go straight to starting a race, you can try changing the line 2162 from `ChangeStateTo(Staging);` to `ChangeStateTo(PreStart);`, and copy lines 2306 & 2307 from the `Staging` state to this `StartRaceMenu` code block.

So currently the block of code handling a start race key press in the `StartRaceMenu` state looks like this:
````
          // Laps or Timed Race Selected
          case 'A': case 'B': case 'C': {
            endLap = raceLaps;
            countingDown = false;
            if (key == 'A') raceType = Standard;
            else if (key == 'B') {raceType = Timed; countingDown = true;}
            else if (key == 'C') {raceType = Drag; endLap = 1;}
            ChangeStateTo(Staging);
          }
````
Try changing it to:
```
            // New Race Type Selected
            case 'A': case 'B': case 'C': {
              endLap = raceLaps;
              countingDown = false;
              if (key == 'A') raceType = Standard;
              else if (key == 'B') {raceType = Timed; countingDown = true;}
              else if (key == 'C') {raceType = Drag; endLap = 1;}
              ChangeStateTo(PreStart);
              preStartTimerDrag = (curMillis % DRAG_PRESTART_RNDM) + DRAG_PRESTART_CNTDWN_BASE;
              newRace = true;
            }
```

Just doing that, might get funky for drag racing because a staging step is how heats are reset and could make drag starts tricky, but you can update the change to only skip the staging step for circuit racing by doing this instead.
```
            // New Race Type Selected
            case 'A': case 'B': case 'C': {
              endLap = raceLaps;
              countingDown = false;
              if (key == 'A') raceType = Standard;
              else if (key == 'B') {raceType = Timed; countingDown = true;}
              else if (key == 'C') {raceType = Drag; endLap = 1;}
              if (key == 'A' || key=='B') {
                  ChangeStateTo(PreStart);
                  newRace = true;
              } else {
                  ChangeStateTo(Staging);
              }
            }
```

Starting Lap Timing Immediately on race start - essentially that sounds like you want the circuit style race start to be the same as a single sensor, drag race start. I think you can achieve this by editing the new race start setup code in the `Race` state code block (starting at line 2478):
Current code:
```
        if(newRace){
          // make sure running race time is starting from 0
          currentTime[0] = 0;
          // record the ms clock time at race start
          startMillis[0] = curMillis;
          if (raceType == Drag) {
            // set lanes Active on race start and set current lap to 1
            for(byte i = 1; i <= laneCount; i++){
              if(laneEnableStatus[i] == StandBy){
                laneEnableStatus[i] = Active;
                // If using a start and finish line triggers, set lapCount to 0 (or 'false')
                // If using a finish line only, set lapCount to 1 (or 'true')
                lapCount[i] = SINGLE_DRAG_TRIGGER;
                flashStatus[i] = 1;
                startMillis[i] = curMillis;
              }
            }
          }
          newRace = false;
        } else { // if newRace = false, returning from a restart.
          startMillis[0] = curMillis - lastXMillis[0][2];
        }
``` 
Try changing to:
```
        if(newRace){
          // make sure running race time is starting from 0
          currentTime[0] = 0;
          // record the ms clock time at race start
          startMillis[0] = curMillis;
          // set lanes Active on race start and set current lap to 1
          for(byte i = 1; i <= laneCount; i++){
            if(laneEnableStatus[i] == StandBy){
              laneEnableStatus[i] = Active;
              flashStatus[i] = 1;
              startMillis[i] = curMillis;
            }
            if (raceType == Drag) {
                // If using a start and finish line triggers, set lapCount to 0 (or 'false')
                // If using a finish line only, set lapCount to 1 (or 'true')
                lapCount[i] = SINGLE_DRAG_TRIGGER;
            } else {   
                lapCount[i] = 1;
            }
          }
          newRace = false;
        } else { // if newRace = false, returning from a restart.
          startMillis[0] = curMillis - lastXMillis[0][2];
        }
```
Note that, of course, once you start editing the code, the line #'s referenced here may not line up exactly anymore, but you should be able to recognize the referenced part nearby, if it has shifted.

Again, I'm suggesting these changes completely blind to the outcome as I don't have hardware to test with right now, but give them a shot and see if they get the results you are interested in.

**Comment by ftsaci** [comment]
Thank you very much for your interest. I think I did not explain myself fully about starting a fast race. What I mean is, plug it in. Do nothing else. Just press the start button and the standard race will start as in the default settings.its possible ? 
Regarding the type counting on the first trigger, yes, it started recording laps on the first trigger. But it records lap times randomly or in a way that I can't understand the order of. It's like 3 minutes when it was 3 seconds.
Thank you so much again for help 😊😊

**Comment by micbay** [comment]
If you want it be ready to run on bootup with no menu inputs you could change the startup to boot straight into the Pre-Staging menu for whatever racetype you want to use, which will be inherently in a state where then pressing the `start` button will start a race.

Starting with the original code, not the modified state skipping code from before, in the `setup()` function change lines 1904 & 1905 from initiating in the main menu state, to start in the `Staging` state, and also add the code from the `StartRaceMenu` switch case (line 2139), for the key press, of the race type you want to boot to automatically

current portion of the `setup()` at lines 1904:
```
  // Set initial state to Menu and initial menu to MainMenu
  ChangeStateTo(Menu);
  currentMenu = MainMenu;
```
to something like:
```
  ChangeStateTo(Staging);
  endLap = raceLaps;
  countingDown = false;
  // These lines that start with double slashes are comments and not executed during runtime
  // you can enable and disable certain code by adding or removing the slashes
  // un-comment the parts you want to try, only pick one of the 3 types at a time
  // for a normal race to X laps
  raceType = Standard;

  // If you want to set it up for a Timed race uncomment these
  // raceType = Timed;
  // countingDown = true;

  // If you want drag racing uncomment these
  // raceType = Drag;
  // endLap = 1;
```

The issue will be what do you do when the race is over, you still need to interact with the menu/display to view results and reset the race or you'd also need to modify the code to change how the `start` button and/or `pause` button react from the main menu state such that they will trigger a re-entry into the `Staging` state directly, as is done for the drag raceType.

The controller works by switching between states (Menu, Staging, PreStart, Race, Paused, Fault and Finish). You can distinguish these states by the Switch statement cases with these labels. On bootup the `setup()` function is run once, initilizing whatever we want for our initial state, then the code will continually loop through the code in the main `loop()` function, executing the code in whatever state it is currently in. There is a flag called `entryFlag` that when set to `true`, indicates that state is being entered for the 1st time. This flag is used to run code that we may need to run just once to setup variables and indicators the first time a new state is looped. Similar to how we put initialization for the whole program in `setup()`, we put whatever initialization we need for a particular state in the `if(entryFlag)` conditional.

I've commented almost every line of code, so people can work out what does what even if you don't fully understand the actual code, you can probably get the gist and move parts that you want to run in different orders around as we've kind of done above.

Regarding the timing discrepancy you are seeing, I can't really guess at the issue without more detail and perhaps an image or video so I can see what you are talking about. Is it a format you are maybe misreading or are you actually not seeing any time index until after at least 1min? The format is mm.ss.0 for running lap time, and mm.ss.000 for captured lap times. It does not display empty pre-fixes so if you see only 1 decimal that should be seconds.

**Comment by ftsaci** [comment]
I added this section under the main menu regarding quick start.

case '#':                
              endLap = raceLaps;
              countingDown = false;
               raceType = Standard;
               ChangeStateTo(PreStart);
              preStartTimerDrag = (curMillis % DRAG_PRESTART_RNDM) + DRAG_PRESTART_CNTDWN_BASE;
              newRace = true;
              break;


 Now I can start the race by pressing the button even if it is not visible on the screen. But I think yours is a better solution. Sorry for my mistake, maybe I misunderstood at that moment. Lap times are wrong. But not minutes. I uploaded a short video about it.

https://www.youtube.com/watch?v=CTiduL7_SE4

edit:
ı did some runs. The times increase with each run without rebooting. 5 seconds reached approximately 5 minutes. yes minuter :( And in the following trials, not only the first lane but also the other lanes started to give their first laps incorrectly.

**Reply by micbay** [reply]
You probably don't know what 'Markup' language is, but you can use it in this forum to format code in a more readable form. For example the code you have here you can hilghlight and click the '<>' button to format it in uni-spaced font. This will wrapped the text inside a leading and trailing set of triple ticks. A 'tick' is the key, usually on the top left of keyboard, that looks like a single quote, ', but leaning sideways. You can also do this manually by just typing in the three ticks before and after the text you want to format, or if it's inline with the text you can use a single `tick` around the text to format, and it will format it as 'code'.

also, if you have several lines of code in a switch statement's case, like this, it's a good habit and a bit more clean, and secure, to wrap them in brackets to control the scope and avoid some potential issues of scope bleeding into other cases you may not want. It's not absolutely necessary for a single line or a few explicit variable updates, but something to be aware of when you start to add in functions, conditionals, or new variables, that have their own scope, inside a case.
```
case '#': {
  endLap = raceLaps;
  countingDown = false;
  raceType = Standard;
  ChangeStateTo(PreStart);
  preStartTimerDrag = (curMillis % DRAG_PRESTART_RNDM) + DRAG_PRESTART_CNTDWN_BASE;
  newRace = true;
}
break;
```
Thanks for the video, on first glance it appears that we may not have setup the initial start time properly when jumping straight in. It's probably just pulling a random value that is in the memory space and using that bogus value to calculate the 1st lap time. Am i correct that the wrong time is always, and only the 1st lap? I'll look into that theory in another reply.

**Comment by ftsaci** [comment]
I'm pretty new here and to the coding. Sorry about that and thank you for your advice. Actually it was just the first round for each lane. But now all the rounds are showing up wrong. I'll flash it again and try again.
By the way, how can we add the start button to the small fast start  code I added? I tried a few times but failed.

**Reply by micbay** [reply]
In the code you posted for `case '#':` you are jumping to the `PreStart` state rather than the `Staging` state. Are you still using this code? I don't think that is skipping the timing initialization, but it might be. So far I'm not seeing why the timing is off if you are starting in the `Staging` state. Perhaps the `entryFlag` is not being set to `true` when transitioning to the `PreStart` state, make sure it is.

**Reply by ftsaci** [reply]
> In the code you posted for `case '#':` you are jumping to the `PreStart` state rather than the `Staging` state. Are you still using this code? I don't think that is skipping the timing initialization, but it might be. So far I'm not seeing why the timing is off if you are starting in the `Staging` state. Perhaps the `entryFlag` is not being set to `true` when transitioning to the `PreStart` state, make sure it is.

Yes, right now, only the small code I sent you is loaded. The lap counting on the first trigger and the prerace screen opening at the beginning are not added. Now the lap times are measured correctly.

**Comment by micbay** [comment]
Yeah, all the code i gave you in my first reply is for using menu keys still to start a race, but skipping the `Staging` step, which isn't what you wanted, and it jumped to the `PreStart` state without setting the `entryFlag` to true which is probably a problem.

I'm not totally clear on where your code is at. Let's try this, start from the original code and implement just this part that updates lines 1904 & 1905 in the `setup()` to jump straight to the `Staging` screen:
```
  ChangeStateTo(Staging);
  endLap = raceLaps;
  countingDown = false;
  // These lines that start with double slashes are comments and not executed during runtime
  // you can enable and disable certain code by adding or removing the slashes
  // un-comment the parts you want to try, only pick one of the 3 types at a time
  // for a normal race to X laps
  raceType = Standard;

  // If you want to set it up for a Timed race uncomment these
  // raceType = Timed;
  // countingDown = true;

  // If you want drag racing uncomment these
  // raceType = Drag;
  // endLap = 1;
```
If that is working as expected then let's try to implement the first lap timing on race start by adding in this updated replacement, to the `if (newRace)` conditional code at line 2478:
```
        if(newRace){
          // make sure running race time is starting from 0
          currentTime[0] = 0;
          // record the ms clock time at race start
          startMillis[0] = curMillis;
          // set lanes Active on race start and set current lap to 1
          for(byte i = 1; i <= laneCount; i++){
            if(laneEnableStatus[i] == StandBy){
              laneEnableStatus[i] = Active;
              flashStatus[i] = 1;
              startMillis[i] = curMillis;
            }
            if (raceType == Drag) {
                // If using a start and finish line triggers, set lapCount to 0 (or 'false')
                // If using a finish line only, set lapCount to 1 (or 'true')
                lapCount[i] = SINGLE_DRAG_TRIGGER;
            } else {   
                lapCount[i] = 1;
            }
          }
          newRace = false;
        } else { // if newRace = false, returning from a restart.
          startMillis[0] = curMillis - lastXMillis[0][2];
        }
```

and let me know how that works.

**Reply by ftsaci** [reply]
I made the first change. The start screen was ready for the race and everything was fine. Then I made the 2nd change for the first trigger lap time. Unfortunately, there was no improvement. The first laps of all lanes are random. The later laps were correct. As I continue the runs without rebooting, the first lap time starts to increase. The first run is like 7 seconds instead of 2 seconds. Later, it increases to 50 seconds, 1 minute, 5 minutes.

**Reply by micbay** [reply]
Just to be clear, after making the initial change so that the system boots directly to the staging screen. You also checked the race timing? At that point, the timing doesn't start until after the first trigger for each lane, but on the 2nd trigger the times for lap 1, and subsequent laps are correct?

But then when you implement the 2nd part that is when the timing issues show up on the 1st lap? It sounds like the time it is using is an absolute time since the boot up and not calculated from the race start time.

**Reply by ftsaci** [reply]
Yes, in the first edit, everything was as it should be. the race begins. The first trigger is taken. On the 2nd trigger, the lap time starts to be written. now try new one

**Comment by micbay** [comment]
For drag racing the time isn't calculated with the same variables as circuit racing because it doesn't need to keep track of multiple laps so I think the problem is that in this case, we are bypassing the `StandBy` lane status, on race start, and setting them `Active` outside of the `ISR()` interrupt function, that is run when a lane is triggered. So we are skipping the initialization of the initial lap start time, that is used to calculate the circuit race lap times.

Try this version where I've added a line `lastXMillis[i][0] = curMillis;` that comes from the line 1284, `lastXMillis[ laneNum ][0] = logMillis;` in the `ISR (PCINT_VECT)` that records the clock time of the triggers that are used to calculate lap times.
```
        if(newRace){
          // make sure running race time is starting from 0
          currentTime[0] = 0;
          // record the ms clock time at race start
          startMillis[0] = curMillis;
          // set lanes Active on race start and set current lap to 1
          for(byte i = 1; i <= laneCount; i++){
            if(laneEnableStatus[i] == StandBy){
              laneEnableStatus[i] = Active;
              flashStatus[i] = 1;
              startMillis[i] = curMillis;
              lastXMillis[i][0] = curMillis;
            }
            if (raceType == Drag) {
                // If using a start and finish line triggers, set lapCount to 0 (or 'false')
                // If using a finish line only, set lapCount to 1 (or 'true')
                lapCount[i] = SINGLE_DRAG_TRIGGER;
            } else {   
                lapCount[i] = 1;
            }
          }
          newRace = false;
        } else { // if newRace = false, returning from a restart.
          startMillis[0] = curMillis - lastXMillis[0][2];
        }
```

**Reply by ftsaci** [reply]
This is great my friend. At the beginning, the race screen greets you and the laps start counting from the first trigger. And all of them are calculated correctly. Thank you very much. For tolerating me for so long. I have other questions but I'll bother you about them later. You've been trying too hard for me today. thank yo so much

**Reply by micbay** [reply]
Cool, glad we got something of what you were looking for, working. Feel free to start as many discussions as you'd like.
